### ISSUE TITLE
[Angular Elements Grid] Row pinning does not work when binding to custom data items.

### ISSUE THREAD

**Issue by MayaKirova** [IgniteUI/igniteui-angular#13036]
## Description  
Row pinning with custom data items seems to pin all rows as if they have the same primary key.

 * igniteui-angular version: 
 * browser: 

## Steps to reproduce  

1. Observe the sample at: https://www.infragistics.com/products/ignite-ui-web-components/web-components/components/grids/grid/row-pinning#web-components-grid-row-pinning-example
or run the attached sample.

3. Try to pin a single row from the grid via the UI.

## Result  
All rows get pinned. Scrollbar disappears.

## Expected result  
A single row should get pinned.

## Attachments  
Note that binding to a simple json work ok, but binding to the custom array of items does not work:

```
            // does not work
            grid.data = new CustomersData();

            // works
            //grid.data = nwCustomersWithOrders;

```

[sample.zip](https://github.com/IgniteUI/igniteui-angular/files/11529586/sample.zip)

The same does not happen for the angular grid: https://stackblitz.com/edit/kva78b

**Comment by damyanpetev** [comment]
> The same does not happen for the angular grid: https://stackblitz.com/edit/kva78b

Ah, sure it will. Just need to downgrade to ES5 output and bring down the build tooling back enough or mess with the plugins, which CBA to repro.


### Excruciating details:
The issue is not really with the component, but with the sample setup. Perhaps not obvious, but creating an extended array that self-populates on creation like this in combination with Webpack build that compiles down to what looks like ES5 (?!) monstrosity:
```ts
class CustomersData extends Array<CustomersDataItem> {
    public constructor() {
        super();
        this.push(new CustomersDataItem(
```
```js
    var CustomersData = /*#__PURE__*/
    function(_Array) {
        (0, _inherits2["default"])(CustomersData, _Array);
        var _super = _createSuper(CustomersData);
        function CustomersData() {
            var _this;
            (0,
            _classCallCheck2["default"])(this, CustomersData);
            _this = _super.call(this);
            _this.push(new CustomersDataItem({
            
            //...
            return _this;
        }
        return (0, _createClass2["default"])(getCustomersData);
```
is a pretty bad idea. For comparison from the Angular Stackblitz from above even when targeting ES5:
```js
var CustomersData = /** @class */function (_super) {
  (0,tslib__WEBPACK_IMPORTED_MODULE_0__.__extends)(CustomersData, _super);
  function CustomersData() {
    var _this = _super.call(this) || this;
    _this.push(new CustomersDataItem({
    
    //...
    return _this;
```
Spot the return/supper handling difference? Not sure what's causing the WC setup to emit ES5-like code (not the config target) and could also be an old bug in `@babel/runtime` or `@babel/plugin-transform-classes`. Like I said, CBA.
Point being, it's basically creating a non-functional read-only collection type.
See, when you `filter` for example, the returned result is a new array instance populated with the results, right? Except, without ES6 handling, the `ctor` populates and you get the exact same result every time:
![image](https://github.com/IgniteUI/igniteui-angular/assets/3198469/3f1c727f-c088-4169-9727-0f500b7666e0)
Same applies for `map()`, `sort`, etc, etc.
You can imagine how that can mess up the pinning, right?

https://github.com/IgniteUI/igniteui-angular/blob/74e2cb780c241fdb8915601812a0b5a9cebdd3d1/projects/igniteui-angular/src/lib/grids/common/pipes.ts#L308-L309

https://github.com/IgniteUI/igniteui-angular/blob/74e2cb780c241fdb8915601812a0b5a9cebdd3d1/projects/igniteui-angular/src/lib/grids/common/pipes.ts#L320-L321

The grid filters the records into two collections - pinned and unpinned, except it ends up with both collections containing the full data which likely throws the whole thing into a disarray (pun intended). No wonder the scrollbar disappears, technically all records are pinned lol 😂 

### TL;DR
The current WC Webpack setup produces broken code for those `extends Array` classes and **_will_** break multiple component features, not just pinning. Perhaps updating the config and/or build tooling could resolve that, but honestly just not using such no-value-added classes is just as good (if not cleaner) solution - a simple `const customerData = [ ... ]` or `getCustomerData` function do the same thing.

Tagging @gmurray81 @HUSSAR-mtrela since this can only be fixed in samples setup/generation output. Feel free to transfer this to a more appropriate repo or create a separate issue and close this one.

**Comment by damyanpetev** [comment]
NB! Actually forgot to check the native behavior (🤦‍♂️):
![image](https://github.com/IgniteUI/igniteui-angular/assets/3198469/518f96ba-708b-4eda-b15b-8ccba58688fb)

So, errm, weirdly that might be the correct behavior and depending on the handling of builders, might be best if we avoid such data classes entirely.

**Comment by damyanpetev** [comment]
> > The same does not happen for the angular grid: https://stackblitz.com/edit/kva78b
> 
> Ah, sure it will. Just need to downgrade to ES5 output and bring down the build tooling back enough or mess with the plugins, which CBA to repro.

Oh I stand corrected, completely missed the fact that sample has batch editing on, so even in ES6 (and per prev comment) it will break without it and the transactions pipe cloning first thing. Here's the repro with same behavior:
https://stackblitz.com/edit/kva78b-wxssnm

**Comment by MayaKirova** [comment]
@dkamburov @damyanpetev Logged the issue for the editor generated code here: https://infragistics.visualstudio.com/NetAdvantage/_workitems/edit/25265

Should we do some temporary workaround on the sample level to get it to a working state? For example copy the data in a new array and assign it to the grid on some event, like init or rendered. Or should we leave it be and wait for a fix in the editor?

**Comment by gmurray81** [comment]
I'll take a look at this today

**Comment by gmurray81** [comment]
I made a fix related to an editor feature we can use to resolve this and messaged Maya and Martin about how we can apply it. I believe we should be able to have these samples resolved soon.
